src/shell/commands/disk_rebalance.cpp

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#include "shell/commands.h"
+#include "shell/argh.h"
+
+#include <math.h>
+#include <fmt/ostream.h>
+#include <dsn/utility/errors.h>
+#include <dsn/utility/output_utils.h>
+#include <dsn/utility/string_conv.h>
+#include <dsn/dist/replication/duplication_common.h>
+
+bool validate_cmd(const argh::parser &cmd,
+                  const std::set<std::string> &params,
+                  const std::set<std::string> &flags)
+{
+    if (cmd.size() > 1) {
+        fmt::print(stderr, "too many params!\n");
+        return false;
+    }
+
+    for (const auto &param : cmd.params()) {
+        if (params.find(param.first) == params.end()) {
+            fmt::print(stderr, "unknown param {} = {}\n", param.first, param.second);
+            return false;
+        }
+    }
+
+    for (const auto &flag : cmd.flags()) {
+        if (params.find(flag) != params.end()) {
+            fmt::print(stderr, "missing value of {}\n", flag);
+            return false;
+        }
+
+        if (flags.find(flag) == flags.end()) {
+            fmt::print(stderr, "unknown flag {}\n", flag);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+std::ostream *get_out_stream(const std::string &file_name)
+{
+    std::ostream *os_ptr = nullptr;
+    if (file_name.empty()) {
+        os_ptr = &std::cout;
+    } else {
+        os_ptr = new std::ofstream(file_name);
+        if (!*os_ptr) {
+            fmt::print(stderr, "open output file {} failed\n", file_name);
+            delete os_ptr;
+        }
+    }
+    return os_ptr;
+}
+
+bool query_disk_info(
+    shell_context *sc,
+    const argh::parser &cmd,
+    const std::string &node_address,
+    const std::string &app_name,
+    /*out*/ std::map<dsn::rpc_address, dsn::error_with<query_disk_info_response>> &err_resps)
+{
+    std::map<dsn::rpc_address, dsn::replication::node_status::type> nodes;
+    auto error = sc->ddl_client->list_nodes(::dsn::replication::node_status::NS_INVALID, nodes);
+    if (error != dsn::ERR_OK) {
+        fmt::print(stderr, "list nodes failed, error={}\n", error.to_string());
+        return false;
+    }
+
+    std::vector<dsn::rpc_address> targets;
+    for (const auto &node : nodes) {
+        if (!node_address.empty() || node_address == node.first.to_std_string()) {
+            targets.push_back(node.first);
```

Comment:
empty node_address is allowed here? Why?

Reply:
if `node_address` is empty, means query `all nodes`.